### PR TITLE
Improve FreeBSD CPUInfo support

### PR DESCRIPTION
**Pull request by rozhuk-im**
## Description
Add CPU temperature reading.
Enable neon detection.

This patch from FreeBSD ports.

## Motivation and Context
Now on all FreeBSD hw platforms that support cpu temp sensor it will be available for user.
#16098 will require to include this feature.

## How Has This Been Tested?
I build and run kodi with this patch, it show CPU temp in info.
Tested on FreeBSD 11.2, 12.1 amd64
Other FreeBSD users use this patch during few years.
Other code not changed.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by rozhuk-im** [comment]
```
TEST(TestCPUInfo, getTemperature)
{
  TemporarySetting command(CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_cpuTempCmd, "echo '50 c'");
  CTemperature t;
  EXPECT_TRUE(g_cpuInfo.getTemperature(t));
  EXPECT_TRUE(t.IsValid());
}
```
test failed probably because cpu driver not loaded on test server, and code cant read actual temp.
What I should do with this?

**Comment by lrusak** [comment]
Thanks for the PR, however, I'd rather this wait until we finally merge #16098

File: xbmc/platform/freebsd/CPUInfoFreebsd.cpp
Comment (Rechi) [inline xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: > Use spaces as tab policy with an indentation size of 2.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

**Review by Rechi — DISMISSED** [review]
Please adjust the commit title from `Improve FreeBSD support` to `Improve FreeBSD CPUInfo support` (same as PR title) so one knows what actually gets improved.

This breaks `TestCPUInfo.GetTemperature` test. That has to be fixed before the PR can be merged.

**Review by Dboii1998 — COMMENTED** [review]
Yes

**Review by Dboii1998 — APPROVED** [review]

**Comment by rozhuk-im** [comment]
Is something else needed here?

**Comment by t6** [comment]
Seems fine to me. Can it be merged?

File: xbmc/platform/freebsd/CPUInfoFreebsd.cpp
Comment (t6) [inline xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: Issue found by @jbeich downstream (see https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=242688):
> FreeBSD neither has /usr/include/asm/* nor getauxval. See base r324815 for similar API. For examples look at other ports. Note, qemu-user *probably* cannot emulate AT_HWCAP. Note2, NEON is always enabled on FreeBSD armv7.
```
xbmc/platform/freebsd/CPUInfoFreebsd.cpp:18:10: fatal error: 'asm/hwcap.h' file not found
#include <asm/hwcap.h>
         ^~~~~~~~~~~~~
1 error generated.
```

Proposed fix:
```diff
--- xbmc/platform/freebsd/CPUInfoFreebsd.cpp.orig
+++ xbmc/platform/freebsd/CPUInfoFreebsd.cpp
@@ -14,11 +14,10 @@
 #include <array>
 #include <vector>
 
-#if (defined(__arm__) && defined(HAS_NEON)) || defined(__aarch64__)
-#include <asm/hwcap.h>
-#include <sys/auxv.h>
-#elif defined(__i386__) || defined(__x86_64__)
+#if defined(__i386__) || defined(__x86_64__)
 #include <cpuid.h>
+#elif __has_include(<sys/auxv.h>)
+#include <sys/auxv.h>
 #endif
 
 #include <sys/resource.h>
@@ -155,9 +154,15 @@ CCPUInfoFreebsd::CCPUInfoFreebsd()
   }
 #endif
 
-#if defined(HAS_NEON) && defined(__arm__)
-  if (getauxval(AT_HWCAP) & HWCAP_NEON)
+#if defined(HAS_NEON)
+#if defined(__ARM_NEON)
+  m_cpuFeatures |= CPU_FEATURE_NEON;
+#elif __has_include(<sys/auxv.h>)
+  unsigned long hwcap = 0;
+  elf_aux_info(AT_HWCAP, &hwcap, sizeof(hwcap));
+  if (hwcap & HWCAP_NEON)
     m_cpuFeatures |= CPU_FEATURE_NEON;
+#endif
 #endif
 }
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: xbmc/platform/freebsd/CPUInfoFreebsd.cpp
Comment (rozhuk-im) [reply xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: Done, please check it.

Comment (t6) [reply xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: Thanks. LGTM. Also built fine on FreeBSD arm64.

**Review by ghost — APPROVED** [review]

File: xbmc/platform/freebsd/CPUInfoFreebsd.cpp
Comment (Rechi) [inline xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: > Use spaces as tab policy with an indentation size of 2.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

File: xbmc/platform/freebsd/CPUInfoFreebsd.cpp
Comment (rozhuk-im) [reply xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: One space removed.

**Comment by t6** [comment]
@Rechi Is this ok style-wise now? Can it be merged?

**Comment by Rechi** [comment]
code style is fine

**Comment by t6** [comment]
Can someone merge it?

**Comment by fritsch** [comment]
@lrusak you worked with cpuinfo last time - your button

File: xbmc/platform/freebsd/CPUInfoFreebsd.cpp
Comment (lrusak) [inline xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: you should do what we do for linux here https://github.com/xbmc/xbmc/blob/master/xbmc/platform/linux/CPUInfoLinux.cpp#L299

This will allow the test to be run regardless of the hardware.

File: xbmc/utils/test/TestCPUInfo.cpp
Comment (lrusak) [inline xbmc/utils/test/TestCPUInfo.cpp]: Not needed if you follow me other comment

**Review by lrusak — CHANGES_REQUESTED** [review]

File: xbmc/platform/freebsd/CPUInfoFreebsd.cpp
Comment (lrusak) [inline xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: if the temp is returned in kelvin you can just create it from kelvin instead of converting yourself. https://github.com/xbmc/xbmc/blob/master/xbmc/utils/Temperature.h#L34

**Comment by t6** [comment]
@rozhuk-im I've addressed the items raised by @lrusak in https://github.com/xbmc/xbmc/commit/8c0cc9cd17d192832fb511c693badacc973fc896. Can you update this pull request?

File: xbmc/platform/freebsd/CPUInfoFreebsd.cpp
Comment (t6) [reply xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: Sounds good, but an implementation for CreateFromKelvin was missing. I had to add one first.

**Comment by rozhuk-im** [comment]
@t6 Thanks for kicking.
I remove ```f``` postfix to avoid double->float->double conversion.

**Comment by lrusak** [comment]
jenkins build this please

**Comment by t6** [comment]
@lrusak Build was fine. Ok to merge?

File: xbmc/platform/freebsd/CPUInfoFreebsd.cpp
Comment (lrusak) [inline xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: c++ style cast please

File: xbmc/platform/freebsd/CPUInfoFreebsd.cpp
Comment (lrusak) [inline xbmc/platform/freebsd/CPUInfoFreebsd.cpp]: `nullptr`

**Comment by lrusak** [comment]
jenkins build this please

**Comment by rozhuk-im** [comment]
offtopic
I see 2 small issue with this API usage:
1. It updates very slow: ~ every 10-30 seconds
2. I show: 50C, not 50,1C

**Review by lrusak — APPROVED** [review]

**Comment by lrusak** [comment]
Thanks for the hard work! Again, sorry it to so long.

The cpuinfo for temperature and cpu speed is for indication only, it really doesn't need to be fast 😺

**Comment by rozhuk-im** [comment]
Thanks!

PS: 10-30 seconds too long IMHO, it will be nice if it will 1-5 seconds, at leas for temperature.

**Pull request merged**
